README.md
# GNU MCU Eclipse OpenOCD - the build scripts

These are the scripts and additional files required to build the
[GNU MCU Eclipse OpenOCD](https://github.com/gnu-mcu-eclipse/openocd).

This release follows the official [OpenOCD](http://openocd.org).

The build scripts use the 
[xPack Build Box (XBB)](https://github.com/xpack/xpack-build-box), 
a set of elaborate build environments based on GCC 7.4 (Docker containers
for GNU/Linux and Windows or a custom HomeBrew for MacOS).

## Repository URLs

- the [GNU MCU Eclipse OpenOCD](https://github.com/gnu-mcu-eclipse/openocd) 
Git remote URL to clone from is https://github.com/gnu-mcu-eclipse/openocd.git
- the [OpenOCD](https://sourceforge.net/p/openocd/code/) Git remote URL is
https://git.code.sf.net/p/openocd/code

Add a remote named `openocd`, and pull the OpenOCD master → master.

## Download the build scripts repo

The build script is available from GitHub and can be 
[viewed online](https://github.com/gnu-mcu-eclipse/openocd-build/blob/master/scripts/build.sh).

To download it, clone the 
[gnu-mcu-eclipse/openocd-build](https://github.com/gnu-mcu-eclipse/openocd-build) 
Git repo, including submodules. 

```console
$ rm -rf ~/Downloads/openocd-build.git
$ git clone --recurse-submodules https://github.com/gnu-mcu-eclipse/openocd-build.git \
  ~/Downloads/openocd-build.git
```

To use the `develop` branch of the build scripts, use:

```console
$ rm -rf ~/Downloads/openocd-build.git
$ git clone --recurse-submodules -b develop https://github.com/gnu-mcu-eclipse/openocd-build.git \
  ~/Downloads/openocd-build.git
```

## The `Work` folder

The script creates a temporary build `Work/openocd-${version}` folder in 
the user home. Although not recommended, if for any reasons you need to 
change the location of the `Work` folder, 
you can redefine `WORK_FOLDER_PATH` variable before invoking the script.

## Changes

Compared to the original OpenOCD distribution, there should be no 
functional changes.

## How to build distributions

### Prerequisites

The prerequisites are common to all binary builds. Please follow the 
instructions in the separate 
[Prerequisites for building binaries](https://gnu-mcu-eclipse.github.io/developer/build-binaries-prerequisites-xbb/) 
page and return when ready.

### Preload the Docker images

Docker does not require to explicitly download new images, but does this 
automatically at first use.

However, since the images used for this build are relatively large, it 
is recommended to load them explicitly before starting the build:

```console
$ bash ~/Downloads/openocd-build.git/scripts/build.sh preload-images
```

The result should look similar to:

```console
$ docker images
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
ilegeul/centos      6-xbb-v2.1          3644716694e8        2 weeks ago         2.99GB
ilegeul/centos32    6-xbb-v2.1          921d03805e50        2 weeks ago         2.91GB
hello-world         latest              f2a91732366c        2 months ago        1.85kB
```

### Update git repos

To keep the development repository in sync with the original OpenOCD 
repository and the RISC-V repository:

- checkout `master`
- pull from `openocd/master`
- checkout `gnu-mcu-eclipse-dev`
- merge `master`
- add a tag like `v0.10.0-8-20180512` after each public release (mind the 
inner version `-8-`)

### Prepare release

To prepare a new release, first determine the OpenOCD version 
(like `0.10.0`) and update the `scripts/VERSION` file. The format is 
`0.10.0-8`. The fourth digit is the GNU MCU Eclipse release number 
of this version.

Add a new set of definitions in the `scripts/container-build.sh`, with 
the versions of various components.

### Update README.md

If necessary, update the main `README.md` with informations related to the
build. Information related to the content should be updated in 
the `README-<version>.md`.

### Create README-&lt;version&gt;.md

Create a copy of the previous one and update.

### Update CHANGELOG.txt

Check `openocd-build.git/CHANGELOG.txt` and add the new release.

### Build

Although it is perfectly possible to build all binaries in a single step 
on a macOS system, due to Docker specifics, it is faster to build the 
GNU/Linux and Windows binaries on a GNU/Linux system and the macOS binary 
separately.

#### Build the GNU/Linux and Windows binaries

The current platform for GNU/Linux and Windows production builds is an 
Ubuntu 17.10 VirtualBox image running on a macMini with 16 GB of RAM 
and a fast SSD.

Before starting a multi-platform build, check if Docker is started:

```console
$ docker info
```

To build both the 32/64-bit Windows and GNU/Linux versions, use `--all`; 
to build selectively, use `--linux64 --win64` or `--linux32 --win32` 
(GNU/Linux can be built alone; Windows also requires the GNU/Linux build).

```console
$ rm -rf ${HOME}/Work/openocd-*
$ bash ~/Downloads/openocd-build.git/scripts/build.sh --all
```

To build one of the previous versions:

```console
$ RELEASE_VERSION=0.10.0-8 bash ~/Downloads/openocd-build.git/scripts/build.sh --all
```

Several hours later, the output of the build script is a set of 4 files 

CHANGELOG.txt

The changes applied to the OpenOCD master branch, and the releases, are:

[2019-04-09]

- v0.10.0-12-20190422 released

[2019-04-09]

- prepare - v0.10.0-12
- update to latest master from Apr 7th, 2019
- update LIBUSB1_VERSION="1.0.22"
- update LIBFTDI_VERSION="1.4"

[2019-01-18]

- v0.10.0-11-20190118 released
- update to latest master from Jan 16, 2019
- RISC-V specific patches were removed, only upstreamed functionality retained.

[2018-10-20]

- v0.10.0-10-20181020 released
- rerun, to fix the macOS file dates
- update the -bit to singular

[2018-10-16]

- v0.10.0-9-20181016 released
- update to latest master
- update to latest RISC-V
- revert some of the RISC-V patches in the common files

[2018-06-19]

- update to latest RISC-V commits, including semihosting

[2018-06-12]

- use separate README-*.md files
- update to latest commits, which include new semihosting (OpenOCD June 6th, RISC-V June 12th)

[2018-05-12]

- v0.10.0-8-20180512 released
- use new build scripts based on XBB
- update to latest commits (OpenOCD April 27th, RISC-V May 8th)

[2018-01-23]

- v0.10.0-7-20180123 released
- move semihosting code to separate files
- use them in RISC-V and ARM
- add 'arm semihosting_resexit enable' to allow exit() to return

[2018-01-12]

- v0.10.0-6-20180112 released
- update to master from Dec 20
- update to riscv from Dec 29
- remove the patch to hide the CSRs, the new version displays only a limited number of them.
- remove the `remote_bitbang.c` patch, since it compiles ok on mingw-w64
- the SiFive board scripts were upstreamed to the RISC-V fork

[2017-11-10]

- v0.10.0-5-20171110-dev released
- update to master from Oct 2
- update to riscv from Nov 4
- target.c & riscv/riscv-0[13].c: hide the 4096 CSRs from `monitor reg`
- update the SiFive board script files
- revert the risc-v changes in `remote_bitbang.c`, since they break the build on mingw-w64

[2017-10-04]

- v0.10.0-4-20171004-*-dev released
- update to master from Aug 10
- update to riscv fom Oct 2
- gdb_server.c: workaround to gdb errors; disable passing errors back to gdb since this risc-v change breaks other targets.

[2017-08-25]

- v0.10.0-3-20170826-*-dev released
- merge RISC-V tag v20170818
- server.c: fix clang warning in getsockname()

[2017-07-03]

- update build script to use Debian 9 Docker containers

[2017-06-22]

- v0.10.0-2-20170622-1535-dev released
- merge RISC-V tag v20170621 

[2017-06-15]

- move the build specific gnu-mcu-eclipse folder to a separate openocd-build project

[2017-06-12]

- add --enable-riscv and #if BUILD_RISCV

scripts/container-libs-functions-source.sh
# -----------------------------------------------------------------------------
# This file is part of the GNU MCU Eclipse distribution.
#   (https://gnu-mcu-eclipse.github.io)
# Copyright (c) 2019 Liviu Ionescu.
#
# Permission to use, copy, modify, and/or distribute this software 
# for any purpose is hereby granted, under the terms of the MIT license.
# -----------------------------------------------------------------------------

# Helper script used in the second edition of the GNU MCU Eclipse build 
# scripts. As the name implies, it should contain only functions and 
# should be included with 'source' by the container build scripts.

# -----------------------------------------------------------------------------

function do_libusb1()
{
  # https://sourceforge.net/projects/libusb/files/libusb-1.0/

  # 2015-09-14, 1.0.20
  # 2018-03-25, 1.0.22

  LIBUSB1_SRC_FOLDER_NAME="libusb-${LIBUSB1_VERSION}"
  LIBUSB1_FOLDER_NAME="${LIBUSB1_SRC_FOLDER_NAME}"
  local libusb1_archive="${LIBUSB1_SRC_FOLDER_NAME}.tar.bz2"
  local libusb1_url="http://sourceforge.net/projects/libusb/files/libusb-1.0/${LIBUSB1_FOLDER_NAME}/${libusb1_archive}"
  # local libusb1_url="https://github.com/gnu-mcu-eclipse/files/raw/master/libs/${libusb1_archive}"

  local libusb1_stamp_file_path="${INSTALL_FOLDER_PATH}/stamp-libusb1-${LIBUSB1_VERSION}-installed"
  if [ ! -f "${libusb1_stamp_file_path}" ]
  then

    cd "${SOURCES_FOLDER_PATH}"

    download_and_extract "${libusb1_url}" "${libusb1_archive}" \
      "${LIBUSB1_SRC_FOLDER_NAME}"

    (
      mkdir -p "${LIBS_BUILD_FOLDER_PATH}/${LIBUSB1_FOLDER_NAME}"
      cd "${LIBS_BUILD_FOLDER_PATH}/${LIBUSB1_FOLDER_NAME}"

      xbb_activate
      xbb_activate_installed_dev

      if [ "${TARGET_PLATFORM}" == "darwin" ]
      then
        # GCC-7 fails to compile Darwin USB.h:
        # error: too many #pragma options align=reset
        export CC=clang
      fi

      export CFLAGS="${XBB_CFLAGS} -Wno-non-literal-null-conversion -Wno-deprecated-declarations  -Wno-format"
      export CPPFLAGS="${XBB_CPPFLAGS}"
      export LDFLAGS="${XBB_LDFLAGS_LIB}"

      if [ ! -f "config.status" ]
      then 

        (
          echo
          echo "Running libusb1 configure..."

          bash "${SOURCES_FOLDER_PATH}/${LIBUSB1_SRC_FOLDER_NAME}/configure" --help

          # --enable-shared required by libftdi.
          # --enable-static required by libftdi. (odd...)
          bash ${DEBUG} "${SOURCES_FOLDER_PATH}/${LIBUSB1_SRC_FOLDER_NAME}/configure" \
            --prefix="${LIBS_INSTALL_FOLDER_PATH}" \
            \
            --build=${BUILD} \
            --host=${HOST} \
            --target=${TARGET} \
            \
            --enable-shared \
            --enable-static
          
          cp "config.log" "${LOGS_FOLDER_PATH}/config-libusb1-log.txt"
        ) 2>&1 | tee "${LOGS_FOLDER_PATH}/configure-libusb1-output.txt"

      fi

      (
        echo
        echo "Running libusb1 make..."

        # Build. 
        # WARNING: Parallel build fails!
        make 
        if [ "${WITH_STRIP}" == "y" ]
        then
          make install-strip
        else
          make install
        fi
      ) 2>&1 | tee "${LOGS_FOLDER_PATH}/make-libusb1-output.txt"
    )

    touch "${libusb1_stamp_file_path}"

  else
    echo "Library libusb1 already installed."
  fi
}

# Required by GNU/Linux and macOS.
function do_libusb0()
{
  # https://sourceforge.net/projects/libusb/files/libusb-compat-0.1/

  # 2013-05-21, 0.1.5, latest
  
  LIBUSB0_SRC_FOLDER_NAME="libusb-compat-${LIBUSB0_VERSION}"
  LIBUSB0_FOLDER_NAME="${LIBUSB0_SRC_FOLDER_NAME}"
  local libusb0_archive="${LIBUSB0_SRC_FOLDER_NAME}.tar.bz2"
  local libusb0_url="http://sourceforge.net/projects/libusb/files/libusb-compat-0.1/${LIBUSB0_FOLDER_NAME}/${libusb0_archive}"
  # local libusb0_url="https://github.com/gnu-mcu-eclipse/files/raw/master/libs/${libusb0_archive}"

  local libusb0_stamp_file_path="${INSTALL_FOLDER_PATH}/stamp-libusb0-${LIBUSB0_VERSION}-installed"
  if [ ! -f "${libusb0_stamp_file_path}" ]
  then

    cd "${SOURCES_FOLDER_PATH}"

    download_and_extract "${libusb0_url}" "${libusb0_archive}" \
      "${LIBUSB0_SRC_FOLDER_NAME}"

    (
      mkdir -p "${LIBS_BUILD_FOLDER_PATH}/${LIBUSB0_FOLDER_NAME}"
      cd "${LIBS_BUILD_FOLDER_PATH}/${LIBUSB0_FOLDER_NAME}"

      xbb_activate
      xbb_activate_installed_dev

      export CFLAGS="${XBB_CFLAGS}"
      export CPPFLAGS="${XBB_CPPFLAGS}"
      export LDFLAGS="${XBB_LDFLAGS_LIB}"

      if [ ! -f "config.status" ]
      then 

        (
          echo
          echo "Running libusb0 configure..."

          bash "${SOURCES_FOLDER_PATH}/${LIBUSB0_SRC_FOLDER_NAME}/configure" --help

          bash ${DEBUG} "${SOURCES_FOLDER_PATH}/${LIBUSB0_SRC_FOLDER_NAME}/configure" \
            --prefix="${LIBS_INSTALL_FOLDER_PATH}" \
            \
            --build=${BUILD} \
            --host=${HOST} \
            --target=${TARGET} \
            \
            --enable-shared \
            --disable-static 
          
          cp "config.log" "${LOGS_FOLDER_PATH}/config-libusb0-log.txt"
        ) 2>&1 | tee "${LOGS_FOLDER_PATH}/configure-libusb0-output.txt"

      fi

      (
        echo
        echo "Running libusb0 make..."

        # Build.
        make -j ${JOBS}
        if [ "${WITH_STRIP}" == "y" ]
        then
          make install-strip
        else
          make install
        fi
      ) 2>&1 | tee "${LOGS_FOLDER_PATH}/make-libusb0-output.txt"
    )

    touch "${libusb0_stamp_file_path}"

  else
    echo "Library libusb0 already installed."
  fi
}

# Required by Windows.
function do_libusb_w32()
{
  # https://sourceforge.net/projects/libusb-win32/files/libusb-win32-releases/
  # 2012-01-17, 1.2.6.0 
  # LIBUSB_W32_VERSION="1.2.6.0" # +PATCH!

  LIBUSB_W32_PREFIX="libusb-win32"
  LIBUSB_W32="${LIBUSB_W32_PREFIX}-${LIBUSB_W32_VERSION}"

  LIBUSB_W32_SRC_FOLDER_NAME="${LIBUSB_W32_PREFIX}-src-${LIBUSB_W32_VERSION}"
  LIBUSB_W32_FOLDER_NAME="${LIBUSB_W32_SRC_FOLDER_NAME}"
  local libusb_w32_archive="${LIBUSB_W32_SRC_FOLDER_NAME}.zip"
  local linusb_w32_url="http://sourceforge.net/projects/libusb-win32/files/libusb-win32-releases/${LIBUSB_W32_VERSION}/${libusb_w32_archive}"
  # local linusb_w32_url="https://github.com/gnu-mcu-eclipse/files/raw/master/libs/${libusb_w32_archive}"

  local libusb_w32_stamp_file_path="${INSTALL_FOLDER_PATH}/stamp-libusb-w32-${LIBUSB_W32_VERSION}-installed"
  if [ ! -f "${libusb_w32_stamp_file_path}" ]
  then

    cd "${SOURCES_FOLDER_PATH}"

    download_and_extract "${linusb_w32_url}" "${libusb_w32_archive}" \
      "${LIBUSB_W32_SRC_FOLDER_NAME}"

    # Mandatory build in the source folder, so make a local copy.
    rm -rf "${LIBS_BUILD_FOLDER_PATH}/${LIBUSB_W32_FOLDER_NAME}"
    mkdir -p "${LIBS_BUILD_FOLDER_PATH}/${LIBUSB_W32_FOLDER_NAME}"
    cp -r "${SOURCES_FOLDER_PATH}/${LIBUSB_W32_SRC_FOLDER_NAME}"/* \
      "${LIBS_BUILD_FOLDER_PATH}/${LIBUSB_W32_FOLDER_NAME}"

    (
      echo
      echo "Running libusb-win32 make..."

      cd "${LIBS_BUILD_FOLDER_PATH}/${LIBUSB_W32_FOLDER_NAME}"

      xbb_activate
      xbb_activate_installed_dev

      # Patch from:
      # https://gitorious.org/jtag-tools/openocd-mingw-build-scripts

      # The conversions are needed to avoid errors like:
      # 'Hunk #1 FAILED at 31 (different line endings).'
      dos2unix src/install.c
      dos2unix src/install_filter_win.c
      dos2unix src/registry.c

      if [ -f "${BUILD_GIT_PATH}/patches/${LIBUSB_W32_PATCH}" ]
      then
        patch -p0 < "${BUILD_GIT_PATH}/patches/${LIBUSB_W32_PATCH}"
      fi

      # Build.
      (
          export CFLAGS="${XBB_CFLAGS} -Wno-unknown-pragmas -Wno-unused-variable -Wno-pointer-sign -Wno-unused-but-set-variable"
          make \
            host_prefix=${CROSS_COMPILE_PREFIX} \
            host_prefix_x86=i686-w64-mingw32 \
            dll
          
      ) 2>&1 | tee "${LOGS_FOLDER_PATH}/make-libusb-w32-output.txt"

      # Manually install, could not find a make target.
      mkdir -p "${LIBS_INSTALL_FOLDER_PATH}/bin"

      # Skipping it does not remove the reference from openocd, so for the
      # moment it is preserved.
      cp -v "${LIBS_BUILD_FOLDER_PATH}/${LIBUSB_W32_FOLDER_NAME}/libusb0.dll" \
        "${LIBS_INSTALL_FOLDER_PATH}/bin"

      mkdir -p "${LIBS_INSTALL_FOLDER_PATH}/lib"
      cp -v "${LIBS_BUILD_FOLDER_PATH}/${LIBUSB_W32_FOLDER_NAME}/libusb.a" \
        "${LIBS_INSTALL_FOLDER_PATH}/lib"

      mkdir -p "${LIBS_INSTALL_FOLDER_PATH}/lib/pkgconfig"
      sed -e "s|XXX|${LIBS_INSTALL_FOLDER_PATH}|" \
        "${BUILD_GIT_PATH}/pkgconfig/${LIBUSB_W32}.pc" \
        > "${LIBS_INSTALL_FOLDER_PATH}/lib/pkgconfig/libusb.pc"

      mkdir -p "${LIBS_INSTALL_FOLDER_PATH}/include/libusb"
      cp -v "${LIBS_BUILD_FOLDER_PATH}/${LIBUSB_W32_FOLDER_NAME}/src/lusb0_usb.h" \
        "${LIBS_INSTALL_FOLDER_PATH}/include/libusb/usb.h"
    )

    touch "${libusb_w32_stamp_file_path}"

  else
    echo "Library libusb-w32 already installed."
  fi
}

function do_libftdi()
{
  # http://www.intra2net.com/en/developer/libftdi/download.php
  # https://www.intra2net.com/en/developer/libftdi/download/libftdi1-1.4.tar.bz2

  # 1.2 (no date)
  # LIBFTDI_VERSION="1.2" # +PATCH!

  LIBFTDI_SRC_FOLDER_NAME="libftdi1-${LIBFTDI_VERSION}"
  LIBFTDI_FOLDER_NAME="${LIBFTDI_SRC_FOLDER_NAME}"
  local libftdi_archive="${LIBFTDI_SRC_FOLDER_NAME}.tar.bz2"

  libftdi_url="http://www.intra2net.com/en/developer/libftdi/download/${libftdi_archive}"
  # local libftdi_url="https://github.com/gnu-mcu-eclipse/files/raw/master/libs/${libftdi_archive}"

  local libftdi_stamp_file_path="${INSTALL_FOLDER_PATH}/stamp-libftdi-${LIBFTDI_VERSION}-installed"
  if [ ! -f "${libftdi_stamp_file_path}" ]
  then

    cd "${SOURCES_FOLDER_PATH}"

    download_and_extract "${libftdi_url}" "${libftdi_archive}" \
      "${LIBFTDI_SRC_FOLDER_NAME}" \
      "${LIBFTDI_PATCH}"

    (
      mkdir -p "${LIBS_BUILD_FOLDER_PATH}/${LIBFTDI_FOLDER_NAME}"
      cd "${LIBS_BUILD_FOLDER_PATH}/${LIBFTDI_FOLDER_NAME}"

      xbb_activate
      xbb_activate_installed_dev

      export CFLAGS="${XBB_CFLAGS}"
      export CPPFLAGS="${XBB_CPPFLAGS}"
      export LDFLAGS="${XBB_LDFLAGS_LIB}"

      (
        echo
        echo "Running libftdi configure..."
        
        if [ "${TARGET_PLATFORM}" == "win32" ]
        then

          # Configure for Windows.
          cmake \
          -DPKG_CONFIG_EXECUTABLE="${PKG_CONFIG}" \
          -DCMAKE_TOOLCHAIN_FILE="${SOURCES_FOLDER_PATH}/${LIBFTDI_SRC_FOLDER_NAME}/cmake/Toolchain-${CROSS_COMPILE_PREFIX}.cmake" \
          -DCMAKE_INSTALL_PREFIX="${LIBS_INSTALL_FOLDER_PATH}" \
          -DLIBUSB_INCLUDE_DIR="${LIBS_INSTALL_FOLDER_PATH}/include/libusb-1.0" \
          -DLIBUSB_LIBRARIES="${LIBS_INSTALL_FOLDER_PATH}/lib/libusb-1.0.a" \
          -DBUILD_TESTS:BOOL=off \
          -DFTDIPP:BOOL=off \
          -DPYTHON_BINDINGS:BOOL=off \
          -DEXAMPLES:BOOL=off \
          -DDOCUMENTATION:BOOL=off \
          -DFTDI_EEPROM:BOOL=off \
          "${SOURCES_FOLDER_PATH}/${LIBFTDI_SRC_FOLDER_NAME}"

        else

          # Configure for GNU/Linux and macOS.
          cmake \
          -DPKG_CONFIG_EXECUTABLE="${PKG_CONFIG}" \
          -DCMAKE_INSTALL_PREFIX="${LIBS_INSTALL_FOLDER_PATH}" \
          -DBUILD_TESTS:BOOL=off \
          -DFTDIPP:BOOL=off \
          -DPYTHON_BINDINGS:BOOL=off \
          -DEXAMPLES:BOOL=off \
          -DDOCUMENTATION:BOOL=off \
          -DFTDI_EEPROM:BOOL=off \
          "${SOURCES_FOLDER_PATH}/${LIBFTDI_SRC_FOLDER_NAME}"

        fi
      ) 2>&1 | tee "${LOGS_FOLDER_PATH}/configure-libftdi-output.txt"

      (
        echo
        echo "Running libftdi make..."

        # Build.
        make -j ${JOBS}
        make install

      ) 2>&1 | tee "${LOGS_FOLDER_PATH}/make-libftdi-output.txt"
    )

    touch "${libftdi_stamp_file_path}"

  else
    echo "Library libftdi already installed."
  fi
}

function do_libiconv()
{
  # https://www.gnu.org/software/libiconv/
  # https://ftp.gnu.org/pub/gnu/libiconv/
  # https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=libiconv

  # 2011-08-07, "1.14"
  # 2017-02-02, "1.15", latest

  LIBICONV_SRC_FOLDER_NAME="libiconv-${LIBICONV_VERSION}"
  LIBICONV_FOLDER_NAME="${LIBICONV_SRC_FOLDER_NAME}"
  local libiconv_archive="${LIBICONV_SRC_FOLDER_NAME}.tar.gz"
  local libiconv_url="https://ftp.gnu.org/pub/gnu/libiconv/${libiconv_archive}"

  local libiconv_stamp_file_path="${INSTALL_FOLDER_PATH}/stamp-libiconv-${LIBICONV_VERSION}-installed"
  if [ ! -f "${libiconv_stamp_file_path}" ]
  then

    cd "${SOURCES_FOLDER_PATH}"

    download_and_extract "${libiconv_url}" "${libiconv_archive}" \
      "${LIBICONV_SRC_FOLDER_NAME}"

    (
      mkdir -p "${LIBS_BUILD_FOLDER_PATH}/${LIBICONV_FOLDER_NAME}"
      cd "${LIBS_BUILD_FOLDER_PATH}/${LIBICONV_FOLDER_NAME}"

      xbb_activate
      xbb_activate_installed_dev

      # -fgnu89-inline fixes "undefined reference to `aliases2_lookup'"
      #  https://savannah.gnu.org/bugs/?47953
      export CFLAGS="${XBB_CFLAGS} -fgnu89-inline -Wno-tautological-compare -Wno-parentheses-equality -Wno-static-in-inline -Wno-pointer-to-int-cast"
      export CPPFLAGS="${XBB_CPPFLAGS}"
      export LDFLAGS="${XBB_LDFLAGS_LIB}"

      if [ ! -f "config.status" ]
      then 

        (
          echo
          echo "Running libiconv configure..."

          bash "${SOURCES_FOLDER_PATH}/${LIBICONV_SRC_FOLDER_NAME}/configure" --help

          bash ${DEBUG} "${SOURCES_FOLDER_PATH}/${LIBICONV_SRC_FOLDER_NAME}/configure" \
            --prefix="${LIBS_INSTALL_FOLDER_PATH}" \
            \
            --build=${BUILD} \
            --host=${HOST} \
            --target=${TARGET} \
            \
            --enable-shared \
            --disable-static \
            --disable-nls

          cp "config.log" "${LOGS_FOLDER_PATH}/config-libiconv-log.txt"
        ) 2>&1 | tee "${LOGS_FOLDER_PATH}/configure-libiconv-output.txt"

      fi

      (
        echo
        echo "Running libiconv make..."

        # Build.
        make -j ${JOBS}
        if [ "${WITH_STRIP}" == "y" ]
        then
          make install-strip
        else
          make install
        fi
      ) 2>&1 | tee "${LOGS_FOLDER_PATH}/make-libiconv-output.txt"
    )

    touch "${libiconv_stamp_file_path}"

  else
    echo "Library libiconv already installed."
  fi
}

function do_hidapi() 
{
  # https://github.com/signal11/hidapi/downloads

  # Oct 26, 2011, "0.7.0"

  # https://github.com/signal11/hidapi/archive/hidapi-0.8.0-rc1.zip
  # Oct 7, 2013, "0.8.0-rc1", latest

  HIDAPI_SRC_FOLDER_NAME="hidapi-hidapi-${HIDAPI_VERSION}"
  HIDAPI_FOLDER_NAME="${HIDAPI_SRC_FOLDER_NAME}"

  local hidapi_archive="hidapi-${HIDAPI_VERSION}.zip"
  local hidapi_url="https://github.com/signal11/hidapi/archive/${hidapi_archive}"
  # local hidapi_url="https://github.com/gnu-mcu-eclipse/files/raw/master/libs/${hidapi_archive}"

  local hidapi_stamp_file_path="${INSTALL_FOLDER_PATH}/stamp-hidapi-${HIDAPI_VERSION}-installed"
  if [ ! -f "${hidapi_stamp_file_path}" ]
  then

    cd "${SOURCES_FOLDER_PATH}"

    download_and_extract "${hidapi_url}" "${hidapi_archive}" \
      "${HIDAPI_SRC_FOLDER_NAME}"

    # Mandatory build in the source folder, so make a local copy.
    rm -rf "${LIBS_BUILD_FOLDER_PATH}/${HIDAPI_FOLDER_NAME}"
    mkdir -p "${LIBS_BUILD_FOLDER_PATH}/${HIDAPI_FOLDER_NAME}"
    cp -r "${SOURCES_FOLDER_PATH}/${HIDAPI_SRC_FOLDER_NAME}"/* \
      "${LIBS_BUILD_FOLDER_PATH}/${HIDAPI_FOLDER_NAME}"

    (
      cd "${LIBS_BUILD_FOLDER_PATH}/${HIDAPI_FOLDER_NAME}"

      xbb_activate
      xbb_activate_installed_dev

      if [ "${TARGET_PLATFORM}" == "win32" ]
      then

        HIDAPI_OBJECT="hid.o"
        HIDAPI_A="libhid.a"

        cd "${LIBS_BUILD_FOLDER_PATH}/${HIDAPI_FOLDER_NAME}/windows"

        export CFLAGS="${XBB_CFLAGS}"

        make -f Makefile.mingw \
          CC=${CROSS_COMPILE_PREFIX}-gcc \
          "${HIDAPI_OBJECT}"

        # Make just compiles the file. Create the archive and convert it to library.
        # No dynamic/shared libs involved.
        ar -r  libhid.a "${HIDAPI_OBJECT}"
        ${CROSS_COMPILE_PREFIX}-ranlib libhid.a

        mkdir -p "${LIBS_INSTALL_FOLDER_PATH}/lib"
        cp -v libhid.a \
          "${LIBS_INSTALL_FOLDER_PATH}/lib"

        mkdir -p "${LIBS_INSTALL_FOLDER_PATH}/lib/pkgconfig"
        sed -e "s|XXX|${LIBS_INSTALL_FOLDER_PATH}|" \
          "${BUILD_GIT_PATH}/pkgconfig/hidapi-${HIDAPI_VERSION}-windows.pc" \
          > "${LIBS_INSTALL_FOLDER_PATH}/lib/pkgconfig/hidapi.pc"

        mkdir -p "${LIBS_INSTALL_FOLDER_PATH}/include/hidapi"
        cp -v "${SOURCES_FOLDER_PATH}/${HIDAPI_FOLDER_NAME}/hidapi/hidapi.h" \
          "${LIBS_INSTALL_FOLDER_PATH}/include/hidapi"

      elif [ "${TARGET_PLATFORM}" == "linux" -o "${TARGET_PLATFORM}" == "darwin" ]
      then

        if [ "${TARGET_PLATFORM}" == "linux" ]
        then
          do_copy_libudev

          export LIBS="-liconv"
        fi

        echo
        echo "Running hidapi bootstrap..."

        bash ./bootstrap

        export CFLAGS="${XBB_CFLAGS}"
        export CPPFLAGS="${XBB_CPPFLAGS}"
        export LDFLAGS="${XBB_LDFLAGS_LIB}"

        (
          echo
          echo "Running hidapi configure..."

          bash "./configure" --help
        
          bash ${DEBUG} "./configure" \
            --prefix="${LIBS_INSTALL_FOLDER_PATH}" \
            \
            --build=${BUILD} \
            --host=${HOST} \
            --target=${TARGET} \
            \
            --enable-shared \
            --disable-static \
            --disable-testgui
        
          cp "config.log" "${LOGS_FOLDER_PATH}/config-hidapi-log.txt"
        ) 2>&1 | tee "${LOGS_FOLDER_PATH}/configure-hidapi-output.txt"

        (
          echo
          echo "Running hidapi make..."

          # Build.
          make -j ${JOBS}
          if [ "${WITH_STRIP}" == "y" ]
          then
            make install-strip
          else
            make install
          fi
        ) 2>&1 | tee "${LOGS_FOLDER_PATH}/make-hidapi-output.txt"

      fi

      rm -f "${LIBS_INSTALL_FOLDER_PATH}"/lib*/libhidapi-hidraw.la
    )

    touch "${hidapi_stamp_file_path}"

  else
    echo "Library hidapi already installed."
  fi
}

function do_copy_libudev()
{
  cp "/usr/include/libudev.h" "${LIBS_INSTALL_FOLDER_PATH}/include"

  if [ "${TARGET_ARCH}" == "x64" ]
  then
    if [ -f "/usr/lib/x86_64-linux-gnu/libudev.so" ]
    then
      cp "/usr/lib/x86_64-linux-gnu/libudev.so" "${LIBS_INSTALL_FOLDER_PATH}/lib"
      cp "/usr/lib/x86_64-linux-gnu/pkgconfig/libudev.pc" "${LIBS_INSTALL_FOLDER_PATH}/lib/pkgconfig"
    elif [ -f "/lib/x86_64-linux-gnu/libudev.so" ]
    then
      # In Debian 9 the location changed to /lib
      cp "/lib/x86_64-linux-gnu/libudev.so" "${LIBS_INSTALL_FOLDER_PATH}/lib"
      cp "/usr/lib/x86_64-linux-gnu/pkgconfig/libudev.pc" "${LIBS_INSTALL_FOLDER_PATH}/lib/pkgconfig"
    elif [ -f "/usr/lib/libudev.so" ]
    then
      # In ARCH the location is /usr/lib
      cp "/usr/lib/libudev.so" "${LIBS_INSTALL_FOLDER_PATH}/lib"
      cp "/usr/lib/pkgconfig/libudev.pc" "${LIBS_INSTALL_FOLDER_PATH}/lib/pkgconfig"
    elif [ -f "/usr/lib64/libudev.so" ]
    then
      # In CentOS the location is /usr/lib64
      cp "/usr/lib64/libudev.so" "${LIBS_INSTALL_FOLDER_PATH}/lib"
      cp "/usr/lib64/pkgconfig/libudev.pc" "${LIBS_INSTALL_FOLDER_PATH}/lib/pkgconfig"
    else
      echo "No libudev.so; abort."
      exit 1
    fi
  elif [ "${TARGET_ARCH}" == "x32" ] 
  then
    if [ -f "/usr/lib/i386-linux-gnu/libudev.so" ]
    then
      cp "/usr/lib/i386-linux-gnu/libudev.so" "${LIBS_INSTALL_FOLDER_PATH}/lib"
      cp "/usr/lib/i386-linux-gnu/pkgconfig/libudev.pc" "${LIBS_INSTALL_FOLDER_PATH}/lib/pkgconfig"
    elif [ -f "/lib/i386-linux-gnu/libudev.so" ]
    then
      # In Debian 9 the location changed to /lib
      cp "/lib/i386-linux-gnu/libudev.so" "${LIBS_INSTALL_FOLDER_PATH}/lib"
      cp "/usr/lib/i386-linux-gnu/pkgconfig/libudev.pc" "${LIBS_INSTALL_FOLDER_PATH}/lib/pkgconfig"
    elif [ -f "/lib/libudev.so.0" ]
    then
      # In CentOS the location is /lib 
      cp "/lib/libudev.so.0" "${LIBS_INSTALL_FOLDER_PATH}/lib"
      cp "/usr/lib/pkgconfig/libudev.pc" "${LIBS_INSTALL_FOLDER_PATH}/lib/pkgconfig"
    else
      echo "No libudev.so; abort."
      exit 1
    fi
  fi
}
